Refactor to increase reliance on dependency injection

whyvra opened the pull request:
Did some refactoring to increase the reliance on dependency injection. I changed a few constructors in order to be able to inject dependencies into the classes that required them.  

There are no changes for the end user and this should be seamless.

However, this does opens up possibilities where we could configure Photino.Blazor's behavior with extension methods. For example, one could add wwwroot as an EmbeddedResource and have the file embedded in the dll instead of outputting them to the bin folder.

Please let me know if there's anything.

Cheers!

whyvra commented:
@MikeYeager Is there anything I can do to get this PR merged?  

It opens the door to easily replacing the PhysicalFileProvider to something like the EmbeddedFileProvider by increasing the reliance on dependency injection for the internal components of Photino.Blazor.  In the future, we could also add more extension methods to configure Photino.Blazor's behavior.

Please review my PR and give me feedback if you think there's anything I can change and/or improve.

Thank you!

MikeYeager commented:
@tomlm Do you have some time to give this a once-over and do some testing? The changes look fine to us, but we're not in a position to do a lot of testing right now.

whyvra commented:
@MikeYeager @tomlm any updates on this?

I'd really like to build a one-file desktop app using Blazor. This PR is the first step in that direction. By increasing the reliance on DI, I can replace the PhysicalFileProvider with the EmbeddedFileProvider. This would allow be to bundle the wwwroot inside the app.

MikeYeager commented:
@whyvra We are getting close to releasing an update to Photino.NET that makes use of the Kestrel (ASP.NET Core) bundled web server. It's wired up to allow you to embed files as resources AND/OR include them as copied files. There is a little bit more description in issue #97 in Photino.NET (https://github.com/tryphotino/photino.NET/issues/97). It doesn't use dependency injection, but it makes it much easier to use the already available local web server to serve embedded resources and/or local files. Does this solve your problem?

whyvra commented:
@MikeYeager Bundling a whole web server seems a bit excessive for my use case. This PR makes no change to the underlying functionality of Photino.Blazor, it just modifies classes constructors to increase the reliance of DI which would allow to introduce things like extension method to modify Photino.Blazor's behavior.  

This PR doesn't have any of those extensions methods. I'm more than fine maintaining my own set of extension methods (or submitting a PR in the future).But we need to fix the classes that generate instance of other classes in their constructors in order for that to happen

MikeYeager commented:
@whyvra I think we're trying to address a couple of different issues at once. I was pointing out that your example scenario can now handled differently if that's your goal. We're not actually bundling anything as the ASP.NET web server is already part of .NET. We're just taking advantage of it. Separately, we're still concerned that this is a breaking change. We'd like to see a default constructor that calls a backwards compatible initialization method and have someone do some testing to ensure this pull request doesn't break existing apps. While we're generally in favor of your approach, we don't feel comfortable introducing it without ensuring backward compatibility. Thanks for your contribution. There is a lot of good work in there.

whyvra commented:
@MikeYeager Is there anything you'd like for me to add to my PR?

whyvra commented:
Hi @MikeYeager 

I removed the constructor I had added to `PhotinoBlazorApp`, thus restoring the default one. Instead, I use the `Initialize` method to supply values to the properties via dependency injection as the `Initialize` method receives a `IServiceProvider` parameter. I also rebased my debug branch against photino.Blazor/debug branch since you guys had a bunch of changes.

I've tested both sample apps on Windows and Linux and they both work as they did previously. I have a small Photino/Blazor app that I tested against too and it works great too, can even use DI to replace the `PhysicalFileProvider` with a `EmbeddedFileProvider`.

Let me know what you think,

Cheers!

ottodobretsberger commented:
Hello @whyvra ,
We have checked the PR and figured it's a good feature implementation. We merged it into our debug branch.
